daemon/migrator.c

```diff
@@ -0,0 +1,805 @@
+};
+
+static void
+migrator_ctor(migrator_t *self, settings_t *settings)
+{
+    log_info("migrator() created");
+    self->mgr_settings         = settings;
+    self->mgr_state            = MIGRATOR_STATE_UNINITIALIZED;
+    self->mgr_later_id         = 0;
+    self->mgr_queue            = g_queue_new();
+    self->mgr_removal_queue    = g_queue_new();
+    migrator_set_state(self, MIGRATOR_STATE_INITIALIZING);
+}
+
+static void
+migrator_dtor(migrator_t *self)
+{
+    log_info("migrator() deleted");
+    migrator_set_state(self, MIGRATOR_STATE_FINAL);
+
+    change_timer(&self->mgr_later_id, 0);
+
+    g_queue_free(self->mgr_queue),
+        self->mgr_queue = NULL;
+
+    g_queue_free(self->mgr_removal_queue),
+        self->mgr_removal_queue = NULL;
+}
+
+migrator_t *
+migrator_create(settings_t *settings)
+{
+    migrator_t *self = g_malloc0(sizeof *self);
+    migrator_ctor(self, settings);
+    return self;
+}
+
+void
+migrator_delete(migrator_t *self)
+{
+    if( self ) {
+        migrator_dtor(self);
+        g_free(self);
+    }
+}
+
+void
+migrator_delete_at(migrator_t **pself)
+{
+    migrator_delete(*pself), *pself = NULL;
+}
+
+void
+migrator_delete_cb(void *self)
+{
+    migrator_delete(self);
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_ATTRIBUTES
+ * ------------------------------------------------------------------------- */
+
+static settings_t *
+migrator_settings(const migrator_t *self)
+{
+    return self->mgr_settings;
+}
+
+static control_t *
+migrator_control(const migrator_t *self)
+{
+    return settings_control(self->mgr_settings);
+}
+
+static const appinfo_t *
+migrator_appinfo(const migrator_t *self, const gchar *appname)
+{
+    return control_appinfo(migrator_control(self), appname);
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_STM
+ * ------------------------------------------------------------------------- */
+
+static migrator_state_t
+migrator_get_state(const migrator_t *self)
+{
+    return self->mgr_state;
+}
+
+static void
+migrator_enter_state(migrator_t *self)
+{
+    switch( migrator_get_state(self) ) {
+    case MIGRATOR_STATE_UNINITIALIZED:
+        abort();
+        break;
+
+    case MIGRATOR_STATE_INITIALIZING:
+        migrator_build_queue(self);
+        break;
+
+    case MIGRATOR_STATE_MIGRATING:
+        break;
+
+    case MIGRATOR_STATE_MIGRATED:
+        /* Triggers settings save -> dequeues all on removal queue
+         * -> empty removal queue in this state -> go to final state
+         */
+        settings_on_migration_finished(migrator_settings(self));
+        break;
+
+    case MIGRATOR_STATE_FINAL:
+        /* Dequeue without migrating */
+        migrator_clear_queue(self);
+        /* Dequeue without removing */
+        migrator_removal_clear_queue(self);
+        break;
+    }
+}
+
+static void
+migrator_leave_state(migrator_t *self)
+{
+    // TODO: Unnecessary?
```

Comment:
All this does is check final state is never left. I could remove this or keep it for similarity with prompter. Fine either way.